hpx/runtime/serialization/detail/polymorphic_nonintrusive_factory.hpp

```diff
@@ -53,13 +53,36 @@ namespace hpx { namespace serialization { namespace detail
     {
         typedef void (*save_function_type) (output_archive& , const void* base);
         typedef void (*load_function_type) (input_archive& , void* base);
-        typedef void* (*create_function_type) ();
+        typedef void* (*create_function_type) (input_archive&);
 
         save_function_type save_function;
         load_function_type load_function;
         create_function_type create_function;
     };
 
+    template <class T>
+    class constructor_selector
+    {
+    public:
+        static T *create(input_archive& ar)
+        {
+            T *t = new T;
+            load_polymorphic(t, ar, hpx::traits::is_nonintrusive_polymorphic<T>());
+            return t;
+        }
+
+    private:
+        static void load_polymorphic(T *t, input_archive& ar, boost::mpl::true_)
+        {
+            serialize(ar, *t, 0);
+        }
+
+        static void load_polymorphic(T *t, input_archive& ar, boost::mpl::false_)
+        {
+            ar >> *t;
+        }
```

Comment:
Pasting off-line discussion here: they are not the same, sadly. Which is one reason why it took me so long to catch this bug.